### PR TITLE
[4657] Reactify Global App (Groups)

### PR DESCRIPTION
**Pull request by aachavarria**
craftercms/craftercms#4657

### CODE REVIEW

File: ui/app/src/components/GroupEditDialog/styles.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import { createStyles, makeStyles } from '@material-ui/core/styles';
+
+export const useStyles = makeStyles((theme) =>
```
Comment (rart) [inline ui/app/src/components/GroupEditDialog/styles.ts]: export default below

File: ui/app/src/components/GroupsGrid/styles.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import { createStyles, makeStyles } from '@material-ui/core/styles';
+
+export const styles = makeStyles((theme) =>
```
Comment (rart) [inline ui/app/src/components/GroupsGrid/styles.ts:19]: export default below

File: ui/app/src/components/GroupsManagement/GroupsManagement.tsx [hunk 3]
```diff
@@ -31,6 +97,32 @@ export default function GroupsManagement() {
...
       />
+      <ErrorBoundary>
+        <SuspenseWithEmptyState
+          resource={resource}
+          suspenseProps={{
+            fallback: <GroupsGridSkeletonTable numOfItems={limit} />
+          }}
+          withEmptyStateProps={{
+            emptyStateProps: {
+              title: <FormattedMessage id="usersGrid.emptyStateMessage" defaultMessage="No Users Found" />
```
Comment (rart) [inline ui/app/src/components/GroupsManagement/GroupsManagement.tsx]: Did you mean groups?

File: ui/app/src/components/EditGroupDialog/EditGroupDialog.tsx [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+ */
+
+import React from 'react';
+import Group from '../../models/Group';
+import { Dialog } from '@material-ui/core';
+import EditGroupDialogContainer from './EditGroupDialogContainer';
+
+export interface EditGroupDialogProps {
+  open: boolean;
+  group: Group;
```
Comment (rart) [inline ui/app/src/components/EditGroupDialog/EditGroupDialog.tsx]: Isn't this supposed to be optional?

File: ui/app/src/components/TransferList/TransferList.tsx [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+  onCheckAllClicked(items: TransferListItem[], checked: boolean): void;
+}
+
+function TransferListColumn(props: TransferListColumnProps) {
+  const { title, items, onItemClick, checkedList, isAllChecked, onCheckAllClicked, inProgressIds } = props;
+  const classes = useStyles();
+  return (
+    <Paper className={classes.listPaper}>
+      <header className={classes.listHeader}>
+        <Checkbox checked={isAllChecked(items)} onChange={(event) => onCheckAllClicked(items, event.target.checked)} />
```
Comment (rart) [inline ui/app/src/components/TransferList/TransferList.tsx]: Move to own file **and directory**

File: ui/app/src/components/TransferList/styles.ts [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    },
+    listPaper: {
+      flexGrow: 1
+    },
+    list: {
+      height: '310px',
+      overflow: 'auto'
+    }
+  })
+);
```
Comment (rart) [inline ui/app/src/components/TransferList/styles.ts]: Split TransferList vs TransferListColumn styles

File: ui/app/src/components/TransferList/TransferList.tsx [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+
+export interface TransferListObject {
+  title?: ReactNode;
+  items: TransferListItem[];
+}
+
+export interface TransferListItem {
+  id: string;
+  title: string;
+  subTitle?: string;
```
Comment (rart) [inline ui/app/src/components/TransferList/TransferList.tsx]:
```suggestion
  subtitle?: string;
```

File: ui/app/src/components/TransferList/TransferList.tsx [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+
+function TransferListColumn(props: TransferListColumnProps) {
+  const { title, items, onItemClick, checkedList, isAllChecked, onCheckAllClicked, inProgressIds } = props;
+  const classes = useStyles();
+  return (
+    <Paper className={classes.listPaper}>
+      <header className={classes.listHeader}>
+        <Checkbox checked={isAllChecked(items)} onChange={(event) => onCheckAllClicked(items, event.target.checked)} />
+        {title && <Typography color="textSecondary">{title}</Typography>}
+        <SearchBar keyword="" onChange={() => {}} classes={{ root: classes.searchBar }} />
```
Comment (rart) [inline ui/app/src/components/TransferList/TransferList.tsx]: onChange doesn't do anything. the SearchBar doesn't do anything.

File: ui/app/src/services/groups.ts [hunk 9]
```diff
@@ -29,14 +29,14 @@ const paginationDefault = {
 };
 
 export function fetchAll(options?: PaginationOptions): Observable<PagedArray<Group>> {
```
Comment (rart) [inline ui/app/src/services/groups.ts]:
```suggestion
export function fetchAll(options?: Partial<PaginationOptions>): Observable<PagedArray<Group>> {
```

File: ui/app/src/components/TransferList/TransferList.tsx [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+  onTargetListItemsRemoved(items: TransferListItem[]): void;
+}
+
+export interface TransferListObject {
+  title?: ReactNode;
+  items: TransferListItem[];
+}
+
+export interface TransferListItem {
+  id: string;
```
Comment (rart) [inline ui/app/src/components/TransferList/TransferList.tsx]:
```suggestion
  id: string | number;
```

File: ui/app/src/components/TransferList/TransferList.tsx [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+import NavigateBeforeIcon from '@material-ui/icons/NavigateBefore';
+import useStyles from './styles';
+import Typography from '@material-ui/core/Typography';
+import SearchBar from '../Controls/SearchBar';
+import { createLookupTable } from '../../utils/object';
+
+interface TransferListProps {
+  source: TransferListObject;
+  target: TransferListObject;
+  inProgressIds: string[];
```
Comment (rart) [inline ui/app/src/components/TransferList/TransferList.tsx]:
```suggestion
  inProgressIds: Array<string | number>;
```

File: ui/app/src/components/TransferList/TransferList.tsx [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+      <header className={classes.listHeader}>
+        <Checkbox checked={isAllChecked(items)} onChange={(event) => onCheckAllClicked(items, event.target.checked)} />
+        {title && <Typography color="textSecondary">{title}</Typography>}
+        <SearchBar keyword="" onChange={() => {}} classes={{ root: classes.searchBar }} />
+      </header>
+      <List dense component="div" role="list" className={classes.list}>
+        {items.map((item, i) => (
+          <ListItem
+            disabled={inProgressIds.includes(item.id)}
+            key={i}
```
Comment (rart) [inline ui/app/src/components/TransferList/TransferList.tsx]: **Make sure to carry this change when you move component to it's file**
```suggestion
            key={item.id}
```

File: ui/app/src/components/TransferList/TransferList.tsx (hunk 8 above)
Comment (rart) [inline ui/app/src/components/TransferList/TransferList.tsx]: Align SearchBar Right.

File: ui/app/src/components/TransferList/TransferList.tsx [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+}
+
+function TransferListColumn(props: TransferListColumnProps) {
+  const { title, items, onItemClick, checkedList, isAllChecked, onCheckAllClicked, inProgressIds } = props;
+  const classes = useStyles();
+  return (
+    <Paper className={classes.listPaper}>
+      <header className={classes.listHeader}>
+        <Checkbox checked={isAllChecked(items)} onChange={(event) => onCheckAllClicked(items, event.target.checked)} />
+        {title && <Typography color="textSecondary">{title}</Typography>}
```
Comment (rart) [inline ui/app/src/components/TransferList/TransferList.tsx]: Make sure to test with a long title. How does title+searcbar behave in that case?

File: ui/app/src/components/TransferList/TransferList.tsx (hunk 8 above)
Comment (aachavarria) [reply ui/app/src/components/TransferList/TransferList.tsx]: that is true the search is not implemented yet

File: ui/app/src/components/EditGroupDialog/EditGroupDialog.tsx [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import React from 'react';
+import Group from '../../models/Group';
+import { Dialog } from '@material-ui/core';
```
Comment (rart) [inline ui/app/src/components/EditGroupDialog/EditGroupDialog.tsx]:
```suggestion
import Dialog from '@material-ui/core/Dialog';
```

File: ui/app/src/components/TransferListColumn/TransferListColumn.tsx [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+import List from '@material-ui/core/List';
+import ListItem from '@material-ui/core/ListItem';
+import ListItemIcon from '@material-ui/core/ListItemIcon';
+import ListItemText from '@material-ui/core/ListItemText';
+import { TransferListItem } from '../TransferList';
+import CircularProgress from '@material-ui/core/CircularProgress';
+import EmptyState from '../SystemStatus/EmptyState';
+import { FormattedMessage } from 'react-intl';
+
+interface TransferListColumnProps {
```
Comment (rart) [inline ui/app/src/components/TransferListColumn/TransferListColumn.tsx]: export

**Review by rart — CHANGES_REQUESTED** [review]

File: ui/app/src/services/groups.ts [hunk 16]
```diff
@@ -68,18 +68,37 @@ export function update(group: Partial<Group>): Observable<Group> {
...
 
-export function trash(groupId: string): Observable<true> {
+export function trash(groupId: number): Observable<true> {
   return del(`/studio/api/2/groups?id=${groupId}`).pipe(mapTo(true));
 }
 
-export function addUsersToGroup(id: number, users: { ids: string[]; usernames: string[] }): Observable<User[]> {
-  return postJSON(`/studio/api/2/groups/${id}/members`, users).pipe(pluck('response', 'users'));
+export function addUserToGroup(groupId: number, username: string): Observable<User[]> {
+  return postJSON(`/studio/api/2/groups/${groupId}/members`, [username]).pipe(pluck('response', 'users'));
```
Comment (rart) [inline ui/app/src/services/groups.ts]:
```suggestion
export function addUserToGroup(groupId: number, username: string): Observable<User> {
  return addUsersToGroup(groupId, [username]).pipe(pluck(0));
```

File: ui/app/src/services/groups.ts [hunk 17]
```diff
@@ -68,18 +68,37 @@ export function update(group: Partial<Group>): Observable<Group> {
...
+}
+
+export function addUsersToGroup(groupId: number, users: { ids: number[] }): Observable<User[]>;
+export function addUsersToGroup(groupId: number, users: { usernames: string[] }): Observable<User[]>;
+export function addUsersToGroup(groupId: number, users: { ids?: number[]; usernames?: string[] }): Observable<User[]> {
+  return postJSON(`/studio/api/2/groups/${groupId}/members`, users).pipe(pluck('response', 'users'));
 }
```
Comment (rart) [inline ui/app/src/services/groups.ts]:
```suggestion
export function addUsersToGroup(groupId: number, ids: number[]): Observable<User[]>;
export function addUsersToGroup(groupId: number, usernames: string[]): Observable<User[]>;
export function addUsersToGroup(groupId: number, idsOrUsernames: Array<number> | Array<string>): Observable<User[]> {
  return postJSON(`/studio/api/2/groups/${groupId}/members`, {
    [typeof idsOrUsernames[0] === 'string' ? 'usernames' : 'ids']: idsOrUsernames
  }).pipe(pluck('response', 'users'));
}
```

File: ui/app/src/services/groups.ts [hunk 18]
```diff
@@ -68,18 +68,37 @@ export function update(group: Partial<Group>): Observable<Group> {
...
 
-export function deleteUserFromGroup(id: number, userId: number, username: string): Observable<true> {
+export function deleteUserFromGroup(groupId: number, userId: number, username: string): Observable<true> {
   const qs = toQueryString({
     userId,
     username
   });
```
Comment (rart) [inline ui/app/src/services/groups.ts]:
```suggestion
export function deleteUserFromGroup(groupId: number, username: string): Observable<true>;
export function deleteUserFromGroup(groupId: number, userId: number): Observable<true>;
export function deleteUserFromGroup(groupId: number, usernameOrUserId: number | string): Observable<true> {
```

File: ui/app/src/services/groups.ts [hunk 19]
```diff
@@ -68,18 +68,37 @@ export function update(group: Partial<Group>): Observable<Group> {
...
-  return del(`/studio/api/2/groups/${id}/members${qs}`).pipe(mapTo(true));
+  return del(`/studio/api/2/groups/${groupId}/members${qs}`).pipe(mapTo(true));
```
Comment (rart) [inline ui/app/src/services/groups.ts]:
```suggestion
  // @ts-ignore - types are correct, signature to accept either is simply not published by deleteUsersFromGroup
  return deleteUsersFromGroup(groupId, [usernameOrUserId]);
```

File: ui/app/src/services/groups.ts [hunk 20]
```diff
@@ -68,18 +68,37 @@ export function update(group: Partial<Group>): Observable<Group> {
...
+export function deleteUsersFromGroup(
+  groupId: number,
+  user: { ids?: number[]; usernames?: string[] }
+): Observable<true> {
+  const qs = toQueryString({
+    userId: user.ids,
+    username: user.usernames
+  });
+  return del(`/studio/api/2/groups/${groupId}/members${qs}`).pipe(mapTo(true));
 }
```
Comment (rart) [inline ui/app/src/services/groups.ts]:
```suggestion
export function deleteUsersFromGroup(groupId: number, userIds: number[]): Observable<true>;
export function deleteUsersFromGroup(groupId: number, usernames: string[]): Observable<true>;
export function deleteUsersFromGroup(groupId: number, usernamesOrUserIds: number[] | string[]): Observable<true> {
  const qs = toQueryString({
    [typeof usernamesOrUserIds[0] === 'string' ? 'username' : 'userId']: usernamesOrUserIds
  });
  return del(`/studio/api/2/groups/${groupId}/members${qs}`).pipe(mapTo(true));
}
```

**Review by rart — CHANGES_REQUESTED** [review]

File: ui/app/src/components/TransferList/TransferList.tsx [hunk 21]
```diff
@@ -0,0 +1,181 @@
...
+import React, { ReactNode, useCallback, useMemo, useState } from 'react';
+import IconButton from '@material-ui/core/IconButton';
+import NavigateNextIcon from '@material-ui/icons/NavigateNext';
+import NavigateBeforeIcon from '@material-ui/icons/NavigateBefore';
+import useStyles from './styles';
+import { createLookupTable } from '../../utils/object';
+import TransferListColumn from '../TransferListColumn';
+import { FormattedMessage } from 'react-intl/dist/react-intl';
+
+interface TransferListProps {
```
Comment (rart) [inline ui/app/src/components/TransferList/TransferList.tsx]: Export

File: ui/app/src/components/GroupsGrid/GroupsGridUI.tsx [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+              </GlobalAppGridCell>
+              <GlobalAppGridCell align="left" className="bordered">
+                <Typography variant="subtitle2">
+                  <FormattedMessage id="words.description" defaultMessage="Description" />
+                </Typography>
+              </GlobalAppGridCell>
+            </GlobalAppGridRow>
+          </TableHead>
+          <TableBody>
+            {groups?.map((group, i) => (
```
Comment (rart) [inline ui/app/src/components/GroupsGrid/GroupsGridUI.tsx]: Why would this ever be null?

File: ui/app/src/components/GroupsManagement/GroupsManagement.tsx [hunk 23]
```diff
@@ -14,23 +14,131 @@
...
       <GlobalAppToolbar
         title={<FormattedMessage id="GlobalMenu.Groups" defaultMessage="Groups" />}
         leftContent={
-          <Button startIcon={<AddIcon />} variant="outlined" color="primary">
+          <Button startIcon={<AddIcon />} variant="outlined" color="primary" onClick={() => setOpenGroupDialog(true)}>
             <FormattedMessage id="sites.createGroup" defaultMessage="Create Group" />
           </Button>
         }
       />
+      <ErrorBoundary>
```
Comment (rart) [inline ui/app/src/components/GroupsManagement/GroupsManagement.tsx]: Remove

**Review by rart — APPROVED** [review]

**Review by rart — APPROVED** [review]

**Pull request merged**